### PR TITLE
WIP: Class-based platform-suites from experiments/platform-suites over 5.7.0-M1

### PR DESCRIPTION
**Pull request by nicerloop**
## Overview

I ported experiments/platform-suites over 5.7.0-M1.

See #744
Replaces #2267 

I still need guidance to improve this proposal to hopefully get it integrated.

---

I hereby agree to the terms of the [JUnit Contributor License Agreement](https://github.com/junit-team/junit5/blob/002a0052926ddee57cf90580fa49bc37e5a72427/CONTRIBUTING.md#junit-contributor-license-agreement).

---

### Definition of Done

- [ ] There are no TODOs left in the code
- [ ] Method [preconditions](https://junit.org/junit5/docs/snapshot/api/org/junit/platform/commons/util/Preconditions.html) are checked and documented in the method's Javadoc
- [ ] [Coding conventions](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#coding-conventions) (e.g. for logging) have been followed
- [ ] Change is covered by [automated tests](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#tests) including corner cases, errors, and exception handling
- [ ] Public API has [Javadoc](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#javadoc) and [`@API` annotations](https://apiguardian-team.github.io/apiguardian/docs/current/api/org/apiguardian/api/API.html)
- [ ] Change is documented in the [User Guide](https://junit.org/junit5/docs/snapshot/user-guide/) and [Release Notes](https://junit.org/junit5/docs/snapshot/user-guide/#release-notes)
- [x] All [continuous integration builds](https://github.com/junit-team/junit5#continuous-integration-builds) pass

### CODE REVIEW

**Comment by sbrannen** [comment]
Tentatively slated for 5.7 M2 for review

**Comment by swaranga** [comment]
Is there support planned support for state sharing across test classes in the same suite. Something like a @BeforeSuite that initializes some expensive resources once which can then be shared across multiple test classes in the suite?

**Comment by nicerloop** [comment]
Hi, for now my interest is in having multiple suites hosting different configurations of test engine (cucumber-junit-platform-engine in my case). I prototyped it by porting experiments/platform-suites over 5.6 and adding configuration parameters. I now hope to have it integrated, starting over on top of 5.7 M1. Will probably need much more than just polishing :)

**Comment by swaranga** [comment]
Was a decision made suite support will be included in 5.7 M2?

**Comment by marcphilipp** [comment]
> Was a decision made whether suite support will be included in 5.7 M2?

We don't think the spike implementation in this PR is production-ready. Adding another layer to the tree might break existing support in build tools and IDEs so it's not something we want to do without careful consideration.

**Comment by mpkorstanje** [comment]
>  Adding another layer to the tree might break existing support in build tools and IDEs so it's not something we want to do without careful consideration.

Which tools would you take into consideration?

 * Maven
 * Gradle
 * IDEA
 * Eclipse
 * JUnit Platform Console Launcher

Any others?

**Comment by marcphilipp** [comment]
I guess at least also VS Code, NetBeans, Bazel, and Ant.

**Comment by mpkorstanje** [comment]
I've checked a few projects. Nothing breaking so far.

- [X] Maven
- [X] Gradle
  - The reporting on a class that is part of a suite collides with that same class executed by `junit-jupiter` in `build/gradle/reports`. Looks like the reporting is still using JUnit 4 assumptions about top level classes. Excluding `junit-jupiter` makes this go away.
- [X] IDEA 2020.2
  - Rerun fails for tests in suite
  - `@Testable` is missing on `@Suite` so no UI support.
- [ ] Eclipse
- [ ] JUnit Platform Console Launcher
- [ ] VS Code
- [ ] NetBeans
- [ ] Bazel
- [ ] Ant

File: junit-platform-suite-api/src/main/java/org/junit/platform/suite/api/Suite.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+ * @see ExcludeTags
+ * @see IncludeEngines
+ * @see ExcludeEngines
+ */
+@Retention(RetentionPolicy.RUNTIME)
+@Target(ElementType.TYPE)
+@Inherited
+@Documented
+@API(status = EXPERIMENTAL, since = "1.7")
+public @interface Suite {
```
Comment (mpkorstanje) [inline junit-platform-suite-api/src/main/java/org/junit/platform/suite/api/Suite.java:44]: Should also be annotated with `@Testable`

File: junit-platform-launcher/src/main/java/org/junit/platform/launcher/core/LauncherDiscoveryResult.java [hunk 2]
```diff
@@ -33,11 +35,13 @@
 
 	private final Map<TestEngine, TestDescriptor> testEngineDescriptors;
 	private final ConfigurationParameters configurationParameters;
+	private final Optional<TestDescriptor> parentDescriptor;
```
Comment (mpkorstanje) [inline junit-platform-launcher/src/main/java/org/junit/platform/launcher/core/LauncherDiscoveryResult.java:38]: Optional should not be used as a wrapper for fields. Consider pass in a null value in the constructor value and wrap it in the getter. As al alternative to passing an null value consider creating multiple constructors.

**Review by mpkorstanje — COMMENTED** [review]
Looking at the code there are quite a few todo items that still need to be resolved.

However I think the biggest obstacle is the way tests suites are discovered. Currently the `junit-platform-launcher` sideloads the suite discovery request in `EngineDiscoveryOrchestrator.collectRequests`.

This comes with a two deficiencies:
 * `junit-platform-launcher` has a dependency on the `junit-platform-suite-api`
 * `suites` is not a `TestEngine` and `suites` can not be excluded/included from discovery like regular engines

Because `suite` is not a test engine I'm in the rather strange situation of having to exclude all other engines so as to avoid having the tests in the suite executed twice. Once by the discovery request from `suite` and once from the regular test engine. 

Now I think the JUnit Platform Launcher shouldn't have to know about the Suite API much like it doesn't know about any other test engine APIs. And looking at the implementation in I wonder if it would be possible to create a `junit-platform-suite-engine` module that either depends on the `junit-platform-launcher` or emulates its behaviour.

**Comment by marcphilipp** [comment]
> Now I think the JUnit Platform Launcher shouldn't have to know about the Suite API much like it doesn't know about any other test engine APIs. And looking at the implementation in I wonder if it would be possible to create a `junit-platform-suite-engine` module that either depends on the `junit-platform-launcher` or emulates its behaviour.
> 
> This would at the very least remove the risk of introducing an extra layer in the tree. It may however require opening up the `junit-platform-launcher` or extracting some code from it.

I vaguely remember there was some reason we didn't consider making it a separate engine but I can't put my finger on it right now. I do think that would solve some of the concerns I'm having with this spike implementation.

@nicerloop @mpkorstanje Would this sth. you'd be interested in exploring? I cannot make any guarantees right now that we're going to accept it but it would be good to see how it would look.

**Comment by mpkorstanje** [comment]
I'd at least be interested in exploring the suite test engine as a separate package within in the `junit-platform-launcher`.  I suspect that untangling it from the launcher will require a bit more direction. But we can consider that as needed.

**Comment by nicerloop** [comment]
Hi, I would surely be interested in exploring the engine option, but as I have very little knowledge of JUnit code-base and project beyond what I discovered in this attempt, I would need some guidance to start one the right foot.

**Comment by marcphilipp** [comment]
@mpkorstanje Sounds good!

Closing this PR in favor of exploring a (standalone) engine option.

**Comment by mpkorstanje** [comment]
Cheers! I've implemented a prototype that seems to work for a very simple test case.

https://github.com/mpkorstanje/junit-platform-suite-engine

I'm returning further discussion to:

https://github.com/junit-team/junit5/issues/744

**Pull request closed without merging**